### PR TITLE
Swift 2.1.0 alpha

### PR DESCRIPTION
**Pull request by PeterMis**

### CODE REVIEW

File: lib/Controller/DisplayGroup.php [hunk 1]
```diff
@@ -1736,4 +1749,181 @@ public function command($displayGroupId)
...
+
+        // get an array of assigned media files
+        /** @var Media[] $assignedFiles */
+        $assignedFiles = $this->mediaFactory->getByDisplayGroupId($displayGroupId);
+
+        $copyMembers = $this->getSanitizer()->getCheckbox('copyMembers', 0);
+        $copyTags = $this->getSanitizer()->getCheckbox('copyTags', 0);
+        $copyAssignments = $this->getSanitizer()->getCheckbox('copyAssignments', 0);
+
+        if (!$this->getUser()->checkEditable($displayGroup)) {
```
Comment (dasgarner) [inline lib/Controller/DisplayGroup.php]: We could make this more efficient if we did the permissions check before we queried for the rest of the data

File: lib/Controller/DisplayGroup.php [hunk 2]
```diff
@@ -1736,4 +1749,181 @@ public function command($displayGroupId)
...
+     * )
+     */
+    public function copy($displayGroupId)
+    {
+        // get display group object
+        /** @var \Xibo\Entity\DisplayGroup $displayGroup */
+        $displayGroup = $this->displayGroupFactory->getById($displayGroupId);
+
+        // get an array of assigned displays
+        /** @var Display[] $membersDisplays */
```
Comment (dasgarner) [inline lib/Controller/DisplayGroup.php]: These type hints shouldn't be necessary if the Factory has the correct `@returns` docblock on it

File: db/migrations/20190315134628_add_bandwidth_limit_column_to_displaygroup_migration.php [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+class AddBandwidthLimitColumnToDisplaygroupMigration extends AbstractMigration
+{
+    /** @inheritdoc */
+    public function change()
+    {
+        $displayGroupTable = $this->table('displaygroup');
+
+        if (!$displayGroupTable->hasColumn('bandwidthLimit')) {
+            $displayGroupTable
+                ->addColumn('bandwidthLimit', 'integer', ['default' => 0, 'null' => true])
```
Comment (dasgarner) [inline db/migrations/20190315134628_add_bandwidth_limit_column_to_displaygroup_migration.php]: are we sure that we need the field nullable if the default value is 0? if so, do we handle a mixture of nulls and 0's in this field when we query for it?

File: lib/Entity/DisplayGroup.php [hunk 4]
```diff
@@ -768,6 +780,7 @@ private function edit()
             'isDynamic' => $this->isDynamic,
             'dynamicCriteria' => $this->dynamicCriteria,
             'dynamicCriteriaTags' => $this->dynamicCriteriaTags,
+            'bandwidthLimit' => $this->bandwidthLimit,
```
Comment (dasgarner) [inline lib/Entity/DisplayGroup.php]: depending on the answer to the migration question, we might need to protect this from null's

File: lib/Helper/XiboUploadHandler.php [hunk 5]
```diff
@@ -50,17 +50,30 @@ protected function handle_form_data($file, $index)
 
             // Get some parameters
             if ($index === null) {
-                if (isset($_REQUEST['name']))
+                if (isset($_REQUEST['name'])) {
```
Comment (dasgarner) [inline lib/Helper/XiboUploadHandler.php]: great!

File: lib/Xmds/Soap.php [hunk 6]
```diff
@@ -1846,19 +1846,24 @@ protected function getIp()
      * Check we haven't exceeded the bandwidth limits
      *  - Note, display logging doesn't work in here, this is CMS level logging
      *
+     * @param int $displayId The Display ID
      * @return bool true if the check passes, false if it fails
+     * @throws NotFoundException
      */
-    protected function checkBandwidth()
+    protected function checkBandwidth($displayId = 0)
```
Comment (dasgarner) [inline lib/Xmds/Soap.php]: Do we ever need to call this without providing a display?

File: lib/Xmds/Soap.php [hunk 7]
```diff
@@ -1846,19 +1846,24 @@ protected function getIp()
...
     {
         // Uncomment to enable auditing.
         //$this->logProcessor->setDisplay(0, true);
 
+        $display = $this->displayFactory->getById($displayId);
```
Comment (dasgarner) [inline lib/Xmds/Soap.php]: Can we use `$this->display` ?

File: views/authed.twig [hunk 8]
```diff
@@ -241,10 +241,16 @@
             <td>
                 <span class="preview"></span>
             </td>
-            <td class="title">
+            <td class="title" title="Name">
```
Comment (dasgarner) [inline views/authed.twig]: Translations are probably needed here, somehow...

File: views/display-form-edit.twig [hunk 9]
```diff
@@ -187,6 +187,23 @@
                         {% endif %}
                         {{ forms.datetime("auditingUntil", title, auditing, helpText) }}
 
+                        {% set title %}{% trans "Bandwidth limit" %}{% endset %}
+                        {% set helpText %}{% trans "The bandwdith limit that should be applied. Enter 0 for no limit." %}{% endset %}
```
Comment (dasgarner) [inline views/display-form-edit.twig]: bandwdith

File: views/authed.twig (hunk 8 above)
Comment (PeterMis) [reply views/authed.twig]: I'm not sure how to do that there 🤔

**Comment by PeterMis** [comment]
Retest this please

**Pull request merged**